### DISCUSSION TITLE
Using more than one target/ST-Link

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#2259] [General]
For my project I have two boards with STM32 MCUs that I'd like to have connected simultaneously.
However, I have not found out how I can decide to which board a binary is downloaded...
I am currently using Sloeber, but I think that with the Arduino IDE it was also not clear which ST-Link is used, and there is no possibility to choose one.
Is this scenario not supported?

**Comment by fpistm** [comment]
Yes it is not supported using stlink.

**Reply by PerennialNovice** [reply]
Ah, so no need trying further...
Any plans on adding such functionality in the future?

**Reply by fpistm** [reply]
That would be nice but not easy to support.
The main idea would be to use the STLink ID and link it to a board name.
Maybe this https://arduino.github.io/arduino-cli/0.19/pluggable-discovery-specification/#board-identification could be used to adress this feature but no time to deeply investigate this, sorry.
Any contribution welcome.

**Comment by PerennialNovice** [comment]
We found a workaround using Windows "pnputil"
it can be used to enable/disable specific USB-devices in a batch file
launching it **as admin!!** before uploading binaries to the board selects the correct ST-Link
On windows, the hardware-ID can be found in device manager => right click on desired debugger => properties => details => Device Instance Path

example:

```
@echo off
REM Replace the following hardware IDs with those of your ST-Link devices
set FIRST_DEVICE_ST_LINK_DEBUGGER_HARDWARE_ID="USB\VID_0483&PID_3754&MI_00\9&24B098B5&0&0000"
set SECOND_DEVICE_ST_LINK_DEBUGGER_HARDWARE_ID="USB\VID_0483&PID_3754&MI_00\9&22D9135C&0&0000"
set THIRD_DEVICE_ST_LINK_DEBUGGER_HARDWARE_ID="USB\VID_0483&PID_3754&MI_00\8&12130BF6&0&0000"

echo Enabling first device ST-Link Debugger
pnputil /enable-device %FIRST_DEVICE_ST_LINK_DEBUGGER_HARDWARE_ID%

REM echo Enabling second device ST-Link Debugger
REM pnputil /enable-device %SECOND_DEVICE_ST_LINK_DEBUGGER_HARDWARE_ID%

REM echo Enabling third device ST-Link Debugger
REM pnputil /enable-device %THIRD_DEVICE_ST_LINK_DEBUGGER_HARDWARE_ID%

REM echo Disabling first device ST-Link Debugger
REM pnputil /disable-device %FIRST_DEVICE_ST_LINK_DEBUGGER_HARDWARE_ID%

echo Disabling second device ST-Link Debugger
pnputil /disable-device %SECOND_DEVICE_ST_LINK_DEBUGGER_HARDWARE_ID%

echo Disabling third device ST-Link Debugger
pnputil /disable-device %THIRD_DEVICE_ST_LINK_DEBUGGER_HARDWARE_ID%

exit
```

I am not familiar with Arduino IDE 1 and 2, so I don't know if its possible there, but using sloeber/eclipse one can define pre- and post-build scripts which can be used to automate the selection project-specific, which probably is what you want...

Following [this instruction](https://windowsreport.com/do-you-want-to-allow-this-app-to-make-changes-to-your-device/#h-2-set-up-a-desktop-shortcut-that-opens-the-program-with-elevated-privileges), you can create shortcuts in your projects source-code folder for the scheduler tasks of the batch scripts. This is necessary, because admin privileges are needed and we don't want to type in credentials or confirm a nagging "allow blahblah to make changes to your computer" screeens.

They can then be launched from within a prebuild_device_X.bat (that can hold further steps to add before the build)

example:

```
@echo off
REM we have to move one step back in path, since inside eclipse, this file is executed from within the build folder,
REM which will make the script fail when called from the command line though...
REM .lnk extension is mandatory!
CALL ..\enable_device_1.lnk
REM CALL ..\someOtherCommand.bat
REM CALL ..\yetAnotherCommand.bat
exit
```

Maybe it is of use to someone :)
